failsafe/failsafe.py

```diff
@@ -81,23 +82,71 @@ async def run(self, callable, *args, **kwargs):
                     raise CircuitOpen() from recent_exception
             try:
                 context.attempts += 1
-                result = await callable(*args, **kwargs)
+                result = await coroutine(*args, **kwargs)
                 self.circuit_breaker.record_success()
                 return result
 
             except Exception as e:
                 if self.retry_policy.should_abort(e):
                     logger.debug("Aborting Failsafe, exception {}".format(type(e).__name__))
+                    _safe_call(self.retry_policy.on_abort)
                     raise
                 recent_exception = e
                 context.errors += 1
                 retry, wait_for = self.retry_policy.should_retry(context, e)
                 self.circuit_breaker.record_failure()
+                _safe_call(self.retry_policy.on_failed_attempt)
 
                 if retry:
                     if wait_for:
                         logger.debug("Waiting {}".format(wait_for))
                         await asyncio.sleep(wait_for)
                     logger.debug("Retrying call")
+                    _safe_call(self.retry_policy.on_retry)
 
+        _safe_call(self.retry_policy.on_retries_exceeded)
         raise RetriesExhausted() from recent_exception
+
+
+class Sync:
+    """
+    Convenience decorator class to use a `Failsafe` instance from "classic" code (as in not using `async`/`await`
+    syntax). Uses `asyncio.run()` (available in python 3.7+) or `asyncio.get_event_loop().run_until_complete()` if
+    `asyncio.run()` isn't available.
+
+    Usage:
+        ```
+        def classic_callable():
+            ...
+
+        failsafe = Sync(Failsafe(retry_policy=...))
+        failsafe.run(classic_callable)
+        ```
+    """
+    def __init__(self, failsafe):
+        self._failsafe = failsafe
+
+    async def _sync_call(self, callable, *args, **kwargs):
+        return callable(*args, **kwargs)
+
+    async def _failsafe_run(self, callable, *args, **kwargs):
+        return await self._failsafe.run(self._sync_call, callable, *args, **kwargs)
+
+    def run(self, callable, *args, **kwargs):
+        """
+        Calls the given method according to the retry_policy and the circuit_breaker
+        specified in the decorated Failsafe instance.
+
+        :param callable: method to call.
+        :param *args:    The original positional arguments of the method to call (<callable>).
+        :param **kwargs: The original keyword arguments of the method to call (<callable>).
+
+        :raises: RetriesExhausted when the retry policy attempts has been reached.
+        :raises: CircuitOpen when the circuit_breaker policy has reached the
+            maximum allowed number of failures
+        """
+        coroutine = self._failsafe_run(callable, *args, **kwargs)
+        if hasattr(asyncio, 'run'):
+            return asyncio.run(coroutine)
```

Comment:
have you researched how other packages/libraries in the community do the sync/async compatibility?

Reply:
Also I think this had some edge case where if you are running asyncio.run code and then inside you have a sync function and try to do this it will break? not sure, I remember some edge case for this but my memory is bad sry :/

Reply:
> have you researched how other packages/libraries in the community do the sync/async compatibility?

I've done some googling, but never found a "proper" solution, libraries either go sync or async. If you know of any good examples I'll happily take a look at them!

> Also I think this had some edge case where if you are running asyncio.run code and then inside you have a sync function and try to do this it will break? not sure, I remember some edge case for this but my memory is bad sry :/

Yeah, more than an edge case, that's a limitation this approach has. If there's already a running loop, such as this:
```python

def sync_function():
  do_stuff()

async def async_coro():
  Sync(Failsafe()).run(sync_function)

asyncio.run(async_coro())
```
You'd get `RuntimeError: asyncio.run() cannot be called from a running event loop` when attempting to execute `asyncio.run()` for the second time inside `Sync.run()`.

The purpose of the `Sync()` class is to offer a convenience method to use Failsafe from non async code, but with caveats. For instance, if you're developing a library, it's not a good idea to use `Sync` there, because if your lib is used from async code, it will blow up. I guess I could do a better work at explaining the caveats in the docs.